chainstore.go

```diff
@@ -1,147 +1,231 @@
 func New(stores ...Store) Store {
-	return &Chain{stores, false}
-	// TODO: make the chain..
-	// call Open(), but in case of error..?
+	c := &Chain{
+		stores: make([]*storeWrapper, 0, len(stores)),
+	}
+	for _, s := range stores {
+		c.stores = append(c.stores, &storeWrapper{Store: s})
+	}
+	return c
 }
 
-func Async(stores ...Store) Store {
-	return &Chain{stores, true}
-}
+// Open opens all the stores.
+func (c *Chain) Open() error {
 
-func (c *Chain) Open() (err error) {
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, c *Chain, i int) {
+			s := c.stores[i]
+
+			s.setErr(s.Open())
+
+			wg.Done()
+		}(&wg, c, i)
+	}
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Close() (err error) {
-	for _, s := range c.stores {
-		err = s.Close()
-		// TODO: we shouldn't stop on first error.. should keep trying to close
-		// and record errors separately
-		if err != nil {
-			return
-		}
+// Close closes all the stores.
+func (c *Chain) Close() error {
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Close failed due to a previous error: %q", err)
+	}
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(wg *sync.WaitGroup, s *storeWrapper) {
+			s.setErr(s.Close())
+			wg.Done()
+		}(&wg, c.stores[i])
 	}
-	return
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Put(key string, val []byte) (err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Put(ctx context.Context, key string, val []byte) (err error) {
+	if !isValidKey(key) {
 		return ErrInvalidKey
 	}
 
-	fn := func() (err error) {
-		for _, s := range c.stores {
-			err = s.Put(key, val)
-			if err != nil {
-				return
-			}
-		}
-		return
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	if c.async {
-		go fn()
-	} else {
-		err = fn()
+
+	fn := func(s Store) error {
+		return s.Put(ctx, key, val)
 	}
-	return
+
+	return c.doWithContext(ctx, fn)
 }
 
-func (c *Chain) Get(key string) (val []byte, err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Get(ctx context.Context, key string) (val []byte, err error) {
+	if !isValidKey(key) {
 		return nil, ErrInvalidKey
 	}
 
-	for i, s := range c.stores {
-		val, err = s.Get(key)
-		if err != nil {
-			return
-		}
+	if err := c.firstErr(); err != nil {
+		return nil, fmt.Errorf("Open failed due to a previous error: %q", err)
+	}
 
-		if len(val) > 0 {
-			if i > 0 {
-				// put the value in all other stores up the chain
-				fn := func() {
-					for n := i - 1; n >= 0; n-- {
-						c.stores[n].Put(key, val) // errors..?
-					}
-				}
-				// if c.async { } else { } ....?
-				go fn()
+	errCh := make(chan error, 1)
+
+	go func() {
```

Comment:
So you create a buffered channel and write to it even before something is reading from it and at the end we get rid of the WaitGroup, that's cool!. I need to test it tho, it seems like the channel will be closed by the time we try to read it so maybe we'll need to add another goroutine somewhere.

Also, what happens with `putBack` if the value is never found and what happens with it if the stores reply with something like [not found, found, not found, not found]?

Good point on the sequential loop, I also thought about it not being sequential but I realized that that would ping all the stores indistinctly, for this particular case I think we actually want them to be sequential and we'll always stop when the first value is found, in that case we will send the value to all stores that don't have it (which are all the stores that we already passed without finding the value). This sequential is not perfect either because we can end up with something like [found, not found, not found, not found].

Maybe we should think a bit on this `Get` again, what do think @pkieltyka?